cmd/clusterctl/pkg/log/logger.go

```diff
@@ -0,0 +1,278 @@
+
+Nb.the clusterctl library also support usage of other loggers.
+
+Following logging conventions are used in clusterctl:
+
+Message
+
+All message should start with a capital letter and finish without the dot.
+
+Log level:
+
+Use Level 0 for the most important user feedback only, e.g.
+- reporting command progress for long running actions
+- reporting command results when required
+
+Use logging Levels 1..4 for providing more details about the command's internal workflow.
+N.B. For sake of simplicity, usage of level 2..4 should be limited as much as possible.
+
+Use logging Levels 5..9 for for providing all the information required for debug purposes/problem investigation.
+N.B. For sake of simplicity, usage of level 6..9 should be limited as much as possible.
+
+Logging WithValues:
+
+Logging WithValues should be preferred to embedding values into error messages because it allows
+machine readability.
+
+Variables name should start with a capital letter.
+
+Logging WithNames:
+
+Logging WithNames should be used carefully.
+Please consider that practices like prefixing the logs with something indicating which part of code
+is generating the log entry might be useful for developers, but it can create confusion for
+the end users because it increases the verbosity without providing information the user can understand/take benefit from.
+
+Logging errors:
+
+A proper error management should always be preferred to the usage of log.Error.
+
+*/
+
+package log
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+)
+
+// LogEntry defines the information that can be used for composing a log line.
+type LogEntry struct {
+	// Prefix of the log lines, composed by the hierarchy of log.WithName values.
+	Prefix string
+
+	// Level of the LogEntry.
+	Level int
+
+	// Values of the log lines, composed by the concatenation of log.WithValues and KeyValue pairs passed to lg.Info/log..
+	Values []interface{}
+}
+
+// LogChannel define a channel where LogEntry can be sent.
+type LogChannel chan LogEntry
+
+// NewLogger returns a new instance of the clusterctl logger.
+func NewLogger() logr.Logger {
+	return &Logger{
+		level:  0,
+		prefix: "",
+		values: nil,
+	}
+}
+
+// NewLogger returns a new instance of the clusterctl logger ready for sending log entries to a channel.
+func NewLoggerWithChannel(ch LogChannel) logr.Logger {
+	return &Logger{
+		ch:     ch,
+		level:  0,
+		prefix: "",
+		values: nil,
+	}
+}
+
+// Logger defines a clusterctl friendly logr.Logger
+type Logger struct {
+	ch     LogChannel
+	level  int
+	prefix string
+	values []interface{}
+}
+
+var _ logr.Logger = &Logger{}
+
+// Enabled tests whether this Logger is enabled.
+// NB. This method relies on klog settings.
+func (l *Logger) Enabled() bool {
+	return bool(klog.V(klog.Level(l.level)))
```

Comment:
Could we instead add a `threshold` field to `Logger` and either add that as a parameter to the constructors or add a new `SetThreshold` method? We can't have a klog dependency in here.

Reply:
done